data_safe_haven/external/api/credentials.py

```diff
@@ -0,0 +1,177 @@
+            msg = "Could not interpret input as an Azure authentication token."
+            raise DataSafeHavenValueError(msg) from exc
+
+    @abstractmethod
+    def get_credential(self) -> TokenCredential:
+        """Get a credential provider from the child class."""
+        pass
+
+    def get_token(
+        self,
+        *scopes: str,
+        **kwargs: Any,
+    ) -> AccessToken:
+        # Require at least 10 minutes of remaining validity
+        validity_cutoff = datetime.now(tz=UTC).timestamp() + 10 * 60
+        if not DeferredCredential.token_ or (
+            DeferredCredential.token_.expires_on < validity_cutoff
+        ):
+            # Generate a new token and store it at class-level token
+            DeferredCredential.token_ = self.get_credential().get_token(
+                *scopes, **kwargs
+            )
+        return DeferredCredential.token_
+
+
+class AzureSdkCredential(DeferredCredential):
+    """
+    Credential loader used by AzureSdk
+
+    Uses AzureCliCredential for authentication
+    """
+
+    def __init__(self, scope: str = "https://management.azure.com/.default") -> None:
+        super().__init__(scopes=[scope])
+
+    def get_credential(self) -> TokenCredential:
+        """Get a new AzureCliCredential."""
+        credential = AzureCliCredential(additionally_allowed_tenants=["*"])
+        # Check that we are logged into Azure
+        try:
+            decoded = self.decode_token(credential.get_token(*self.scopes).token)
+            self.logger.info(
+                "You are currently logged into the [blue]Azure CLI[/] with the following details:"
+            )
+            self.logger.info(
+                f"... user: [green]{decoded['name']}[/] ({decoded['oid']})"
+            )
+            self.logger.info(
+                f"... tenant: [green]{decoded['upn'].split('@')[1]}[/] ({decoded['tid']})"
+            )
+        except (CredentialUnavailableError, DataSafeHavenValueError) as exc:
+            self.logger.error(
+                "Please authenticate with Azure: run '[green]az login[/]' using [bold]infrastructure administrator[/] credentials."
+            )
+            msg = "Error getting account information from Azure CLI."
+            raise DataSafeHavenAzureError(msg) from exc
+        return credential
+
+
+class GraphApiCredential(DeferredCredential):
+    """
+    Credential loader used by GraphApi
+
+    Uses DeviceCodeCredential for authentication
+    """
+
+    def __init__(
+        self,
+        tenant_id: str,
+        scopes: Sequence[str] = [],
+    ) -> None:
+        super().__init__(scopes=scopes, tenant_id=tenant_id)
+
+    def get_credential(self) -> TokenCredential:
+        """Get a new DeviceCodeCredential, using cached credentials if they are available"""
+        cache_name = f"dsh-{self.tenant_id}"
+        authentication_record_path = (
+            pathlib.Path.home() / f".msal-authentication-cache-{cache_name}"
```

Comment:
Will this be the correct path for all OSs?

Reply:
Good point. Let's use `config_dir` here.

Reply:
Done in e95a0d7a0